bson/bsoncodec/struct_codec.go

```diff
@@ -86,8 +86,7 @@ func (sc *StructCodec) EncodeValue(r EncodeContext, vw bsonrw.ValueWriter, i int
 		if iz, ok := encoder.(CodecZeroer); ok {
 			iszero = iz.IsTypeZero
 		}
-
-		if desc.omitEmpty && iszero(rv.Interface()) {
+		if (rv.Kind() == reflect.Ptr || desc.omitEmpty) && iszero(rv.Interface()) {
```

Comment:
Hi, I was testing this but I think this line is wrong. If we do not have omitempty and the kind is pointer we want to add the key just with a null value in case the pointer is a nil pointer. Problem here is that iszero does not work with pointers and apparently is defined only for objectid

Reply:
That is correct, we want to write a BSON Null if the pointer is nil, not skip it entirely.

Reply:
So yes this code here is wrong as it is skipping any nil pointer

Reply:
I update the code, now `WriteNull` is called when an nil pointer is encountered.
Should I put it in a standalone function ?

Reply:
For the encode path, we should just let the individual encoders handle it, we don't need to have the struct codec care if it has a `nil` pointer. We should add `nil` pointer handling for all of the  encoders as well as the decoders. This will allow users to create an encoder that handles `nil` pointers differently, e.g. if a user wants to put some default data instead of BSON null.

Reply:
You're right, but I have no idea how to RegisterEncoder for a nil value in the current registry. Maybe we should modify these encoders to add pointer support?